pkg/sshclient/client.go

```diff
@@ -0,0 +1,88 @@
+package sshclient
+
+import (
+	"fmt"
+	"os/exec"
+
+	"github.com/dcos/dcos-core-cli/pkg/pluginutil"
+
+	"github.com/dcos/dcos-cli/api"
+	"github.com/sirupsen/logrus"
+)
+
+// Client to execute commands over SSH.
+type Client struct {
+	BinaryPath string
+	Args       []string
+	cmd        *exec.Cmd
+	Logger     *logrus.Logger
+}
+
+var baseArgs = []string{
+	"-A",
+	"-t",
+}
+
+// NewClient creates a new client to start a SSH session.
+func NewClient(options []string, config, user, proxy, host string) (*Client, error) {
+	c := &Client{}
+	c.Logger = pluginutil.Logger()
+	binPath, err := exec.LookPath("ssh")
+	if err != nil {
+		return nil, err
+	}
+	c.BinaryPath = binPath
+
+	c.configureOptions(options, config)
+	c.configureDestination(proxy, host, config, user)
+
+	return c, nil
+}
+
+func (c *Client) configureOptions(options []string, config string) {
+	for _, option := range options {
+		c.Args = append(c.Args, "-o", fmt.Sprintf("%s", option))
+	}
+
+	if config != "" {
+		c.Args = append(c.Args, "-F", fmt.Sprintf("%s", config))
+	}
+}
+
+func (c *Client) configureUser(config, user string) {
+	if config != "" {
+		return
+	}
+	c.Args = append(c.Args, "-l", fmt.Sprintf("%s", user))
+}
+
+func (c *Client) configureDestination(proxy, host, config, user string) {
+	c.Logger.Debugf("Trying to establish connection to %s\n", host)
+	if proxy != "" {
+		c.Logger.Debugf("Using %s as a proxy node\n", proxy)
+		c.Args = append(c.Args, baseArgs...)
+		c.configureUser(config, user)
+		c.Args = append(c.Args, fmt.Sprintf("%s", proxy))
+		c.Args = append(c.Args, "ssh")
+	}
+
+	c.Args = append(c.Args, baseArgs...)
+	c.configureUser(config, user)
+	c.Args = append(c.Args, fmt.Sprintf("%s", host))
+}
+
+// Run adds the optional remote command and starts the SSH session.
+func (c *Client) Run(ctx api.Context, command []string) error {
+	args := c.Args
+	if len(command) > 0 {
+		args = append(args, command...)
+	}
+	cmd := exec.Command(c.BinaryPath, args...)
+	c.Logger.Debugf("Running: %v\n", cmd.Args)
+
+	cmd.Stdin = ctx.Input()
+	cmd.Stdout = ctx.Out()
+	cmd.Stderr = ctx.ErrOut()
```

Comment:
it'd be more easily testable if these were passed individually, maybe in the constructor, instead of passing the whole context.